url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/lexer/LexerBuilder.java

```diff
@@ -0,0 +1,358 @@
+		lexer.setTerminals(separation.getTokenSymbols());
+	</pre>
+	
+	@see fri.patterns.interpreter.parsergenerator.syntax.builder.SyntaxSeparation
+	@see fri.patterns.interpreter.parsergenerator.lexer.StandardLexerRules
+	@author (c) 2002, Fritz Ritzberger
+*/
+
+public class LexerBuilder
+{
+	protected Map charConsumers;
+	protected List ignoredSymbols;
+	public static boolean DEBUG;	// defaults to false
+
+	/**
+		Creates a LexerBuilder (from lexer rules) that provides a Lexer.
+		@param lexerSyntax lexer rule (without token and ignored, use SyntaxSeparation for that)
+		@param ignoredSymbols list of ignored symbols, NOT enclosed in backquotes!
+	*/
+	public LexerBuilder(Syntax lexerSyntax, List ignoredSymbols)
+		throws LexerException, SyntaxException
+	{
+		this.ignoredSymbols = ignoredSymbols;
+		build(lexerSyntax);
+	}
+
+
+	/** Returns the built Lexer. */
+	public Lexer getLexer()	{
+		return new LexerImpl(ignoredSymbols, charConsumers);
+	}
+		
+	/** Returns the built Lexer, loaded with passed input (file, stream, string, ...). */
+	public Lexer getLexer(Object input)
+		throws IOException
+	{
+		Lexer lexer = getLexer();
+		lexer.setInput(input);
+		return lexer;
+	}
+
+
+	private void build(Syntax lexerSyntax)
+		throws LexerException, SyntaxException
+	{
+		SyntaxSeparation.IntArray deleteIndexes = new SyntaxSeparation.IntArray(lexerSyntax.size());
+		if (DEBUG)
+			System.err.println("Processing lexer rules: \n"+lexerSyntax);
+
+		// resolve scanner rules to Consumers and put it into a hashtable
+		this.charConsumers = new Hashtable(lexerSyntax.size());
+		for (int i = 0; i < lexerSyntax.size(); i++)
+			translateLexerRule(lexerSyntax.getRule(i), i, deleteIndexes);
+		deleteIndexes.removeIndexesFrom(lexerSyntax);
+		
+		// check for unresolved repeatable and nullable rules and delete them from lexer syntax
+		for (int i = 0; i < lexerSyntax.size(); i++)	{
+			Rule rule = lexerSyntax.getRule(i);
+			String nonterm = rule.getNonterminal();
+			if (checkNullableRule(nonterm, rule, i, deleteIndexes) == false)
+				if (checkRepeatableRule(nonterm, rule, i, deleteIndexes) == false)
+					throw new LexerException("Found no character consumer for nullable or repeatable rule "+rule);
+		}
+		deleteIndexes.removeIndexesFrom(lexerSyntax);
+		
+		if (lexerSyntax.size() > 0)	{	// not all rules have been resolved to character consumers
+			throw new LexerException("Could not process rules in lexer syntax: "+lexerSyntax);
+		}
+
+		// resolve all symbolic consumer references after all consumers have been created
+		Map done = new Hashtable();	// beware of recursion
+		for (Iterator it = charConsumers.entrySet().iterator(); it.hasNext(); )	{
+			Consumer cc = (Consumer) ((Map.Entry)it.next()).getValue();
+			cc.resolveConsumerReferences(charConsumers, done);
+		}
+	}
+
+
+	private void translateLexerRule(Rule rule, int index, SyntaxSeparation.IntArray deleteIndexes)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.